hpx/lcos/local/channel.hpp

```diff
@@ -0,0 +1,672 @@
+//  Copyright (c) 2016 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_LCOS_LOCAL_CHANNEL_JUL_23_2016_0707PM)
+#define HPX_LCOS_LOCAL_CHANNEL_JUL_23_2016_0707PM
+
+#include <hpx/config.hpp>
+#include <hpx/exception.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/lcos/local/no_mutex.hpp>
+#include <hpx/lcos/local/receive_buffer.hpp>
+#include <hpx/lcos/local/spinlock.hpp>
+#include <hpx/runtime/launch_policy.hpp>
+#include <hpx/util/assert.hpp>
+#include <hpx/util/atomic_count.hpp>
+#include <hpx/util/iterator_facade.hpp>
+#include <hpx/util/scoped_unlock.hpp>
+#include <hpx/util/unused.hpp>
+
+#include <boost/exception_ptr.hpp>
+#include <boost/intrusive_ptr.hpp>
+
+#include <cstdlib>
+#include <iterator>
+#include <mutex>
+#include <utility>
+
+namespace hpx { namespace lcos { namespace local
+{
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        ///////////////////////////////////////////////////////////////////////
+        template <typename T>
+        struct channel_base
+        {
+            channel_base()
+              : count_(0)
+            {}
+
+            virtual ~channel_base() {}
+
+            virtual hpx::future<T> get(std::size_t generation,
+                bool blocking = false) = 0;
+            virtual bool try_get(std::size_t generation,
+                hpx::future<T>* f = nullptr) = 0;
+            virtual void set(std::size_t generation, T && t) = 0;
+            virtual void close() = 0;
+
+            long use_count() const { return count_; }
+            long addref() { return ++count_; }
+            long release() { return --count_; }
+
+        private:
+            hpx::util::atomic_count count_;
+        };
+
+        template <typename T>
+        void intrusive_ptr_add_ref(channel_base<T>* p)
+        {
+            p->addref();
+        }
+
+        template <typename T>
+        void intrusive_ptr_release(channel_base<T>* p)
+        {
+            if (0 == p->release())
+                delete p;
+        }
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename T>
+        class unlimited_channel : public channel_base<T>
+        {
+            typedef hpx::lcos::local::spinlock mutex_type;
+
+            HPX_NON_COPYABLE(unlimited_channel);
+
+        public:
+            unlimited_channel()
+              : closed_(false)
+            {}
+
+        protected:
+            hpx::future<T> get(std::size_t generation, bool blocking)
```

Comment:
This is an internal argument. It informs the implementation that the caller is blocking which changes the error handling inside the implementation. We might want to revisit this, it's not pretty.